firedrake/petsc.py

```diff
@@ -249,3 +251,46 @@ def inserted_options(self):
         finally:
             for k in self.to_delete:
                 del self.options_object[self.options_prefix + k]
+
+
+def _extract_comm(obj):
+    comm = None
+    if isinstance(obj, (PETSc.Comm, mpi.MPI.Comm)):
+        try:
+            if mpi.is_pyop2_comm(obj):
+                comm = obj
+            else:
+                internal_comm = obj.Get_attr(mpi.innercomm_keyval)
+                if internal_comm is None:
+                    comm = obj
+                else:
+                    comm = internal_comm
+        except mpi.PyOP2CommError:
+            pass
+    elif hasattr(obj, "_comm"):
+        comm = obj._comm
+    elif hasattr(obj, "comm"):
+        comm = obj.comm
+    return comm
+
+
+def garbage_cleanup(obj):
+    """ Cleans up garbage PETSc objects on Firedrake object or any comm
+
+    :arg obj: Any Firedrake object or any comm
+    """
+    gc.collect()
+    comm = _extract_comm(obj)
+    if comm:
```

Comment:
If we fail to find a communicator we should fail loudly here rather than do nothing.

Reply:
I'll emit a warning, but I don't want to raise an exception.

Some users will just ignore warnings though.

Reply:
Why not raise an exception? We are already assuming that users "know what they are doing" if they are calling this.

Reply:
Because raising an exception in parallel can cause a deadlock if the exception is only raised on some ranks.

Despite what [this comment](https://github.com/OP2/PyOP2/blob/master/pyop2/mpi.py#L513) suggests, I have definitely experienced deadlocks as a result of some, but not all, ranks raising an exception!

Reply:
Now raising exception following discussion.